src/gurt/shm_lru.c

```diff
@@ -0,0 +1,475 @@
+	if (node->key_size > LRU_ALLOC_SIZE_THRESHOLD || cache->key_size == 0)
+		shm_free((char *)cache + node->key);
+	if (node->data_size > LRU_ALLOC_SIZE_THRESHOLD || cache->data_size == 0)
+		shm_free((char *)cache + node->data);
+
+	/* updata the linked list in the hash bucket */
+	off_node_to_remove = (long int)node - (long int)cache;
+	if (off_hash[node->idx_bucket] == off_node_to_remove) {
+		/* the head node of this bucket */
+		off_hash[node->idx_bucket] = node->off_hnext;
+	} else {
+		/* loop the chain of nodes in this bucket */
+		offset = off_hash[node->idx_bucket];
+		while (offset) {
+			node_hnext = (shm_lru_node_t *)((long int)cache + (long int)offset);
+			if (node_hnext->off_hnext == off_node_to_remove) {
+				node_hnext->off_hnext = node->off_hnext;
+				node->off_hnext       = 0;
+				break;
+			}
+			offset = node_hnext->off_hnext;
+		}
+	}
+
+	first_av_saved = subcache->first_av;
+	subcache->first_av =
+	    (off_node_to_remove - (long int)subcache->off_nodelist) / sizeof(shm_lru_node_t);
+
+	/* remove from list */
+	if (node->off_prev > 0) {
+		node_prev = (shm_lru_node_t *)((long int)cache + (long int)node->off_prev);
+		node_prev->off_next = node->off_next;
+	}
+	if (node->off_next > 0) {
+		node_next = (shm_lru_node_t *)((long int)cache + (long int)node->off_next);
+		node_next->off_prev = node->off_prev;
+	}
+
+	if (node == node_tail)
+		subcache->off_tail = node->off_prev;
+
+	if (off_node_to_remove == subcache->off_head)
+		subcache->off_head = node->off_next;
+
+	node->off_hnext = first_av_saved;
+	subcache->size--;
+	if (subcache->size == 0)
+		/* cache is empty */
+		subcache->off_head = 0;
+	return 0;
+}
+
+static inline int
+key_cmp(shm_lru_cache_t *cache, shm_lru_node_t *node, void *key, uint32_t key_size)
+{
+	if (node->key_size != key_size)
+		/* key size does not match */
+		return 1;
+
+	return memcmp((char *)cache + node->key, key, key_size);
+}
+
+int
+shm_lru_get(shm_lru_cache_t *cache, void *key, uint32_t key_size, shm_lru_node_t **node_found,
+	    void **val)
+{
+	uint64_t             hash         = d_hash_murmur64(key, key_size, 0);
+	uint32_t             idx_subcache = (uint32_t)((hash >> 32) % cache->n_subcache);
+	shm_lru_cache_var_t *cache_list =
+	    (shm_lru_cache_var_t *)((long int)cache + cache->off_cache_list);
+	int *off_hash =
+	    (int *)((long int)cache + (long int)cache_list[idx_subcache].off_hashbuckets);
+	uint32_t        index  = (uint32_t)(hash % cache->capacity);
+	int             offset = off_hash[index];
+	shm_lru_node_t *node;
+
+	shm_mutex_lock(&cache_list[idx_subcache].lock);
+
+	*node_found = NULL;
+	if (cache->key_size != 0)
+		D_ASSERT(cache->key_size == key_size);
+
+	while (offset) {
+		node = (shm_lru_node_t *)((long int)cache + (long int)offset);
+		if (key_cmp(cache, node, key, key_size) == 0) {
+			lru_move_to_head(cache, &cache_list[idx_subcache], node);
+			*val = ((void *)cache + (long int)node->data);
```

Comment:
i can cause segfault if i pass NULL or invalid val?

Reply:
Now "val" is checked. Thank you!